File: ent/schema/article.go
// ent/schema/article.go

/*
 * @Description:
 * @Author: 安知鱼
 * @Date: 2025-07-25 09:51:07
 * @LastEditTime: 2025-08-13 19:01:58
 * @LastEditors: 安知鱼
 */
package schema

import (
	"time"

	"github.com/anzhiyu-c/anheyu-app/ent/schema/mixin"

	"entgo.io/ent"
	"entgo.io/ent/dialect/entsql"
	"entgo.io/ent/schema"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
)

// Article holds the schema definition for the Article entity.
type Article struct {
	ent.Schema
}

// Annotations of the Article.
func (Article) Annotations() []schema.Annotation {
	return []schema.Annotation{
		entsql.WithComments(true),
		schema.Comment("文章表"),
	}
}

// Mixin of the Article.
func (Article) Mixin() []ent.Mixin {
	return []ent.Mixin{
		mixin.SoftDeleteMixin{},
	}
}

// Fields of the Article.
func (Article) Fields() []ent.Field {
	return []ent.Field{
		// --- 基础字段 ---
		field.Uint("id"),
		field.Uint("owner_id").
			Comment("文章作者ID，关联到users表").
			Default(1), // 默认值为管理员ID(1)
		field.Time("created_at").Default(time.Now),
		field.Time("updated_at").Default(time.Now),
		field.String("title").Comment("文章标题").NotEmpty(),
		field.Text("content_md").Comment("文章的 Markdown 原文").Optional(),
		field.Text("content_html").Comment("由 content_md 解析和净化后的 HTML").Optional(),
		field.String("cover_url").Comment("封面图URL").Optional(),
		field.Enum("status").Values("DRAFT", "PUBLISHED", "ARCHIVED", "SCHEDULED").Default("DRAFT"),
		field.Int("view_count").Comment("浏览次数").Default(0).NonNegative(),
		field.Int("word_count").Comment("总字数").Default(0).NonNegative(),
		field.Int("reading_time").Comment("阅读时长(分钟)").Default(0).NonNegative(),
		field.String("ip_location").Comment("作者IP属地").Optional(),
		field.String("primary_color").
			Comment("主色调，取自 top_img_url 或 cover_url").
			Optional().
			Default("#b4bfe2"),
		field.Bool("is_primary_color_manual").
			Comment("主色调是否为手动设置").
			Default(false),
		field.Bool("show_on_home").
			Comment("是否在首页显示，发布后默认显示在首页").
			Default(true),
		field.Int("home_sort").
			Comment("首页推荐文章排序，0 表示不展示，>0 表示展示，数值越小越靠前").
			Default(0).
			NonNegative(),
		field.Int("pin_sort").
			Comment("置顶排序，0 表示不置顶，>0 表示置顶，数值越小越靠前").
			Default(0).
			NonNegative(),
		field.String("top_img_url").
			Comment("顶部图URL，可选。若不填，则在保存时自动使用封面图URL").
			Optional(),
		field.JSON("summaries", []string{}).
			Comment("文章摘要列表，用于随机摘要功能").
			Optional(),
		field.String("abbrlink").
			Comment("永久链接，用于替换ID，需要保证唯一性").
			Optional().
			Unique().
			Nillable(),
		field.Bool("copyright").
			Comment("是否显示版权信息").
			Default(true),
		field.Bool("is_reprint").
			Comment("是否为转载文章").
			Default(false),
		field.String("copyright_author").
			Comment("版权作者").
			Optional(),
		field.String("copyright_author_href").
			Comment("版权作者链接").
			Optional(),
		field.String("copyright_url").
			Comment("版权来源链接").
			Optional(),
		field.String("keywords").
			Comment("文章关键词，用于SEO优化").
			Optional(),

		// --- 定时发布相关字段 ---
		field.Time("scheduled_at").
			Comment("定时发布时间，当status为SCHEDULED时有效").
			Optional().
			Nillable(),

		// --- 审核相关字段（多人共创功能） ---
		field.Enum("review_status").
			Values("NONE", "PENDING", "APPROVED", "REJECTED").
			Comment("审核状态：NONE-无需审核, PENDING-待审核, APPROVED-已通过, REJECTED-已拒绝").
			Default("NONE"),
		field.String("review_comment").
			Comment("审核意见").
			Optional(),
		field.Time("reviewed_at").
			Comment("审核时间").
			Optional().
			Nillable(),
		field.Uint("reviewed_by").
			Comment("审核人ID").
			Optional().
			Nillable(),

		// --- 下架相关字段（PRO版管理员功能） ---
		field.Bool("is_takedown").
			Comment("是否已下架：下架后前台不显示，后台可见").
			Default(false),
		field.String("takedown_reason").
			Comment("下架原因").
			Optional(),
		field.Time("takedown_at").
			Comment("下架时间").
			Optional().
			Nillable(),
		field.Uint("takedown_by").
			Comment("下架操作人ID").
			Optional().
			Nillable(),

		// --- 扩展配置字段 ---
		field.JSON("extra_config", map[string]interface{}{}).
			Optional().
			Comment("文章扩展配置（JSON格式，用于存储各种可选功能配置，如 enable_ai_podcast 等）"),

		// --- 会员权益相关字段 ---
		field.Bool("exclude_from_membership").
			Comment("是否排除在会员权益外：true表示会员也需要单独购买此文章").
			Default(false),

		// --- 文档模式相关字段 ---
		field.Bool("is_doc").
			Comment("是否为文档模式：文档模式的文章会在文档页面展示").
			Default(false),
		field.Uint("doc_series_id").
			Comment("文档系列ID，关联到doc_series表").
			Optional().
			Nillable(),
		field.Int("doc_sort").
			Comment("文档在系列中的排序，数值越小越靠前").
			Default(0).
			NonNegative(),

		// --- 版权区域按钮显示控制字段 ---
		field.Bool("show_reward_button").
			Comment("是否显示打赏作者按钮").
			Default(true),
		field.Bool("show_share_button").
			Comment("是否显示分享按钮").
			Default(true),
		field.Bool("show_subscribe_button").
			Comment("是否显示订阅按钮").
			Default(true),
	}
}

// Indexes of the Article.
func (Article) Indexes() []ent.Index {
	return []ent.Index{
		// 前台列表查询：已发布+未删除+未下架+审核状态+首页显示
		index.Fields("deleted_at", "status", "is_takedown", "review_status", "show_on_home"),
		// 排序索引
		index.Fields("deleted_at", "status", "pin_sort", "created_at"),
		// 归档查询
		index.Fields("deleted_at", "status", "created_at"),
		// 文档模式查询
		index.Fields("deleted_at", "is_doc", "doc_series_id", "doc_sort"),
		// 作者查询
		index.Fields("deleted_at", "owner_id", "status"),
	}
}

// Edges of the Article.
func (Article) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("post_tags", PostTag.Type),
		edge.To("post_categories", PostCategory.Type),
		edge.To("comments", Comment.Type),
		edge.To("histories", ArticleHistory.Type),
		edge.From("doc_series", DocSeries.Type).
			Ref("articles").
			Field("doc_series_id").
			Unique(),
	}
}

File: frontend/src/types/article.ts
/**
 * 文章相关类型定义（与 anheyu-app 后端 API 保持一致）
 */

// ===================================
//          文章标签 (PostTag)
// ===================================

export interface PostTag {
  id: string;
  created_at: string;
  updated_at: string;
  name: string;
  slug?: string;
  count: number;
}

// ===================================
//          文章分类 (PostCategory)
// ===================================

export interface PostCategory {
  id: string;
  created_at: string;
  updated_at: string;
  name: string;
  slug?: string;
  description: string;
  count: number;
  is_series: boolean;
  sort_order: number;
}

// ===================================
//      混合内容流项 (FeedItem)
// ===================================

/**
 * @description 混合内容项（文章或商品），用于首页文章列表
 */
export interface FeedItem {
  id: string;
  item_type: "article" | "product";
  title: string;
  cover_url: string;
  created_at: string;
  // 文章特有字段
  pin_sort?: number;
  comment_count?: number;
  post_tags?: PostTag[];
  post_categories?: PostCategory[];
  is_doc?: boolean;
  doc_series_id?: string;
  primary_color?: string; // 文章主色调，用于分类标签背景色
  // 商品特有字段
  min_price?: number;
  max_price?: number;
  total_sales?: number;
}

// ===================================
//          API 响应类型
// ===================================

export interface FeedListResponse {
  list: FeedItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface CategoryListResponse {
  list: PostCategory[];
  total: number;
}

export interface ArticleListResponse {
  list: Article[];
  total: number;
}

// ===================================
//          查询参数类型
// ===================================

export interface GetFeedListParams {
  page?: number;
  pageSize?: number;
  category?: string;
  tag?: string;
  year?: number;
  month?: number;
}

export interface GetArticleListParams {
  page?: number;
  pageSize?: number;
  category?: string;
  tag?: string;
  year?: number;
  month?: number;
}

export interface GetCategoryListParams {
  sort?: "count" | "name";
}

// ===================================
//          归档 (Archive)
// ===================================

export interface Archive {
  year: number;
  month: number;
  count: number;
}

// ===================================
//      文章详情 (Article)
// ===================================

/**
 * 相邻文章链接
 */
export interface ArticleLink {
  id: string;
  title: string;
  cover_url?: string;
  abbrlink: string;
  created_at: string;
  primary_color?: string;
  is_doc?: boolean;
  doc_series_id?: string;
}

/**
 * 最近文章（用于侧边栏等组件）
 */
export interface RecentArticle {
  id: number | string;
  title: string;
  abbrlink?: string;
  cover_url?: string;
  created_at: string;
  is_doc?: boolean;
  doc_series_id?: number | string;
}

/**
 * 全文隐藏配置
 */
export interface FullTextHiddenConfig {
  enabled: boolean;
  button_text?: string;
  initial_visible_height?: number;
  is_content_truncated?: boolean;
}

/**
 * 文章扩展配置
 */
export interface ArticleExtraConfig {
  enable_ai_podcast?: boolean;
  enable_toc?: boolean;
  custom_css?: string;
  custom_js?: string;
}

/**
 * 完整文章详情
 */
export interface Article {
  id: string;
  created_at: string;
  updated_at: string;
  title: string;
  abbrlink: string;
  content_html: string;
  content_md?: string;
  cover_url?: string;
  top_img_url?: string;
  summaries?: string[];
  word_count: number;
  reading_time: number;
  view_count: number;
  comment_count?: number;
  is_reprint: boolean;
  copyright_author?: string;
  copyright_author_href?: string;
  copyright_url?: string;
  keywords?: string;
  primary_color?: string;
  ip_location?: string;
  status: string;
  review_status?: string;
  is_preview?: boolean;
  pin_sort?: number;
  post_tags: PostTag[];
  post_categories: PostCategory[];
  // 文档模式相关字段
  is_doc?: boolean;
  doc_series_id?: string | number;
  prev_article?: ArticleLink;
  next_article?: ArticleLink;
  related_articles?: ArticleLink[];
  full_text_hidden_config?: FullTextHiddenConfig;
  extra_config?: ArticleExtraConfig;
  // 版权相关字段
  copyright?: boolean; // 是否显示版权信息
  owner_nickname?: string; // 发布者昵称
  owner_name?: string; // 发布者名称
  owner_avatar?: string; // 发布者头像
  owner_email?: string; // 发布者邮箱（用于与头部一致解析 QQ/Gravatar 头像）
  // 版权区域按钮显示控制（文章级别）
  show_reward_button?: boolean;
  show_share_button?: boolean;
  show_subscribe_button?: boolean;
}
